### PR TITLE
Convert a hard-warning about named static lifetimes into a lint

### LINKED ISSUE #96956: "warning: unnecessary lifetime parameter" unsilenceable lint
Given the following code: 

```rust
pub fn thing<'a>(_s: &'a str) where 'a: 'static { }
```

The current output is:

```
warning: unnecessary lifetime parameter `'a`
 --> src/lib.rs:2:37
  |
2 | pub fn thing<'a>(_s: &'a str) where 'a: 'static { }
  |                                     ^^
  |
  = help: you can use the `'static` lifetime directly, in place of `'a`

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 1.27s
```

Ideally the output should also say something like this:

```
  = note: `#[warn(needless_lifetimes)]` on by default
```

And of course then `#[allow(needless_lifetimes)]` ought to work.

### Versions affected

```
rustcargo@zealot:/volatile/rustcargo/Rustup/Game/server$ rustc -vV
rustc 1.62.0-nightly (ecd44958e 2022-05-10)
binary: rustc
commit-hash: ecd44958e0a21110d09862ee080d95a4ca6c52f8
commit-date: 2022-05-10
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.1
rustcargo@zealot:/volatile/rustcargo/Rustup/Game/server$ 
```

The example code is accepted without any complaint by stable and beta,  However, in both stable and beta, clippy has the same weird name-less complaint.  And saying `#![allow(clippy::needless_lifetimes)]` doesn't suppress the message.

I searched for "lifetime" in the clippy lints db.  Of those that came up, `needless_lifetimes` and `redundant_static_lifetimes` seemed like they might be it, but neither of those suppresses the lint on stable's clippy.

### PR DESCRIPTION
**Pull request by jeremydavis519**
Fixes #96956.

This change is pretty simple, but any feedback is appreciated.

Thanks to cjgillot for the clear instructions in the issue's thread. They made this so much easier. :)

### CODE REVIEW

**Comment by petrochenkov** [comment]
Perhaps a `UNUSED_LIFETIMES` lint can be reused instead?
It fits pretty well, and this case looks too niche for introducing a new separate lint.

File: compiler/rustc_resolve/Cargo.toml [hunk 1]
```diff
@@ -21,6 +21,7 @@ rustc_expand = { path = "../rustc_expand" }
 rustc_feature = { path = "../rustc_feature" }
 rustc_hir = { path = "../rustc_hir" }
 rustc_index = { path = "../rustc_index" }
+rustc_lint_defs = { path = "../rustc_lint_defs" }
```
Comment (petrochenkov) [inline compiler/rustc_resolve/Cargo.toml:24]: `rustc_resolve` typically uses lints through `rustc_session::lint`, so adding this shouldn't be necessary.

**Comment by jeremydavis519** [comment]
I agree that this is a very niche case, but I hesitate to lump this together with `UNUSED_LIFETIMES` because the lifetime isn't necessarily unused in this case. The problem is that it's unnecessary. And, maybe more importantly, this problem has a different solution: replacing the lifetime with `'static` instead of just removing the lifetime.

**Comment by estebank** [comment]
@jeremydavis519 I think @petrochenkov's proposal is to reuse the unused lifetimes name for the lint so that these two are effectively a "family" of errors in the same lint, the user visible output would have the same amount of context as it does today so people wouldn't be confused about what they should do. It would remove some granularity on allowing one and not the other, but that should be... rare.

File: compiler/rustc_resolve/src/late/lifetimes.rs [hunk 2]
```diff
@@ -1255,20 +1256,24 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
...
-                                        lifetime.name.ident(),
-                                    ))
-                                    .emit();
+                                        );
+                                        let help = &format!(
+                                            "you can use the `'static` lifetime directly, in place of `{}`",
+                                            lifetime.name.ident(),
+                                        );
+                                        lint.build(msg)
+                                            .help(help)
```
Comment (estebank) [inline compiler/rustc_resolve/src/late/lifetimes.rs:1273]: Not necessarily something to do on this PR, but we have enough context to turn the `help` into a structured suggestion, by inspecting the bounds and suggesting:

```
LL -     type Y<'a: 'static>;
LL +     type Y<'static>;
   |            ~~~~~~~
```

**Comment by jeremydavis519** [comment]
Alright, I don't mind grouping the lints together. My main concern is that I would be mildly confused if I saw `#[allow(unused_lifetimes)]` but all of the lifetimes were used (even if some were used unnecessarily). But I can't think of any situations where someone would actually selectively allow or deny this particular warning, so it probably doesn't matter.

**Comment by apiraino** [comment]
What's the status here? I'll switch to waiting on author since I think there are pending comments. Feel free to update the status of this PR with `@rustbot ready` when ready, thanks!

@rustbot author

**Comment by mkatychev** [comment]
@jeremydavis519  what's the timeline on moving forward with the proposal? I'd be happy to address feedback

**Comment by mkatychev** [comment]
@estebank PR contributor hasn't been active in this PR for over 2 months, is there a process for taking over the feature ask?

**Comment by estebank** [comment]
@mkatychev you can pull from @jeremydavis519's branch, build upon it and publish a new PR referencing this one.

**Comment by JohnCSimon** [comment]
@jeremydavis519 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**